src/macaron/repo_finder/commit_finder.py

```diff
@@ -0,0 +1,160 @@
+import logging
+import re
+from re import Pattern
+
+from git import TagReference
+from packageurl import PackageURL
+from pydriller import Git
+
+logger: logging.Logger = logging.getLogger(__name__)
+
+ALPHANUMERIC = "0-9a-z"
+PREFIX = "(?:.*)"
+INFIX = "[^0-9]*"
+split_pattern = re.compile(f"[^{ALPHANUMERIC}]", flags=re.IGNORECASE)
+validation_pattern = re.compile(f"[{ALPHANUMERIC}]+", flags=re.IGNORECASE)
+
+
+def get_commit_from_version(git_obj: Git, purl: PackageURL) -> tuple[str, str]:
+    """Try to find the matching commit in a repository of a given version via tags.
+
+    The version of the passed PackageURL is used to match with the tags in the target repository.
+
+    Parameters
+    ----------
+    git_obj: Git
+        The repository.
+    purl: PackageURL | None
+        The PURL of the artifact.
+
+    Returns
+    -------
+    tuple[str, str]
+        The branch name and digest as a tuple.
+    """
+    if purl.version is None:
+        logger.debug("Missing version for artifact: %s", purl.name)
+        return "", ""
+    logger.debug("Searching for commit of artifact version using tags: %s@%s", purl.name, purl.version)
+    # Create version based regex from the PackageURL for matching against tags.
+    # The version is split on non-alphanumeric characters to separate the version parts from the non-version parts.
+    # e.g. 1.2.3-DEV -> [1, 2, 3, DEV]
+    split = split_pattern.split(purl.version)
+    logger.debug("Split version: %s", split)
+    if not split:
+        split = [purl.version]
+    this_version_pattern = ""
+    for part in split:
+        # Validate the split part by checking it is only comprised of alphanumeric characters.
+        valid = validation_pattern.match(part)
+        if not valid:
+            continue
+        if this_version_pattern:
+            # To maximise chances of matching tags, the regex between the version parts allows for zero or more
+            # non-numeric characters.
+            this_version_pattern = this_version_pattern + INFIX
+        this_version_pattern = this_version_pattern + str(part)
+    # Prepend the optional prefix, add a named capture group for the version, and enforce end of string analysis.
+    this_version_pattern = PREFIX + "(?P<version>" + this_version_pattern + ")" + "$"
+    logger.debug("Created pattern: %s", this_version_pattern)
+    target_version_pattern = re.compile(this_version_pattern, flags=re.IGNORECASE)
+
+    # A pattern for very weakly confirming the presence of a version.
+    has_version_pattern = re.compile(".*[0-9].*", flags=re.IGNORECASE)
+
+    # Tags are examined as followed:
+    # - Any without a corresponding commit are discarded.
+    # - If any tag matches the has_name_pattern, only tags that match it will be examined.
+    # - If no tag matches the has_name_pattern, all tags will be examined.
+    named_tags = []
+    other_tags = []
```

Comment:
I think this is not entirely true. At least for mypy, when a generic list is initialized like this, the type of the list is inferred from the first `.append` to the list. You can try running the following examples through mypy.

**Example 1**: mypy cannot infer the type of an empty list just from initialization.
```python
from typing_extensions import reveal_type
a = []
reveal_type(a)
```
`mypy t.py` shows:
```
t.py:2: error: Need type annotation for "a" (hint: "a: List[<type>] = ...")  [var-annotated]
t.py:3: note: Revealed type is "builtins.list[Any]"
Found 1 error in 1 file (checked 1 source file)
```


**Example 2**: mypy can infer the type of the list from the first `.append`.
```python
from typing_extensions import reveal_type
a = []
a.append(1)
reveal_type(a)
```
`mypy t.py` shows:
```
t.py:4: note: Revealed type is "builtins.list[builtins.int]"
Success: no issues found in 1 source file
```

**Example 3**: mypy infers the type of the list from the first `.append` and type-checks the subsequent `.append` accordingly.
```python
from typing_extensions import reveal_type
a = []
a.append(1)
a.append("hello")
reveal_type(a)
```
`mpyp t.py` shows:
```
t.py:4: error: Argument 1 to "append" of "list" has incompatible type "str"; expected "int"  [arg-type]
t.py:5: note: Revealed type is "builtins.list[builtins.int]"
Found 1 error in 1 file (checked 1 source file)
```

This is somewhat unrelated, but I have seen this pattern of type inference for containers from the first push/append in another language, [Rust](https://doc.rust-lang.org/rust-by-example/types/inference.html).

In this particular case, my personal opinion is it is fine whether these lists are type-annotated or not. You can make the argument for both sides. This is not about typing issues but more about code readability and developer ergonomics.

Reply:
Hmm that's true. I think that there might be some issues with my mypy settings in my editor. It didn't raise any errors in example 3 so I thought It couldn't resolve the types from the usages of `append`. This explanation makes sense, Thanks @nathanwn. 
@benmss I guess it's up to you regarding the final decision, I'm also fine with both options as they are not typing issues.

Reply:
Nice conversation. Since these are local variables, the name of the variables can be enough to improve readability. See a similar discussion on usage of `var` in Java: https://openjdk.org/projects/amber/guides/lvti-faq

Reply:
Interesting for sure. I've added the type as I think it does help readability in this case.